lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -786,15 +801,52 @@ listPools spl =
             apparentPerformance
 
 joinStakePool
-    :: forall n k.
+    :: forall ctx s t n k.
         ( PaymentAddress n k
+        , Buildable (ErrValidateSelection t)
+        , s ~ SeqState n k
         , k ~ ShelleyKey
+        , HardDerivation k
+        , ctx ~ ApiLayer s t k
         )
-    => ApiT PoolId
+    => ctx
+    -> StakePoolLayer IO
+    -> ApiT PoolId
     -> ApiT WalletId
     -> ApiWalletPassphrase
     -> Handler (ApiTransaction n)
-joinStakePool _ _ _ = throwError err501
+joinStakePool ctx spl (ApiT poolId) (ApiT wid) passwd = do
+    allPools <- listPools spl
+    case L.find (\(ApiStakePool pId _ _) -> pId == ApiT poolId) allPools of
+        Nothing ->
+            liftHandler $ throwE (ErrCreateCertNoSuchPool poolId)
+        Just _ -> do
+            let (ApiWalletPassphrase (ApiT pwd)) = passwd
+
+            selection <- liftHandler $ withWorkerCtx ctx wid liftE1 $ \wrk ->
+                W.createCert @_ @s @t wrk wid
+
+            (tx, meta, time, wit) <- liftHandler $ withWorkerCtx ctx wid liftE2 $ \wrk ->
+                W.signCert @_ @s @t @k wrk wid () pwd selection poolId
+
+            liftHandler $ withWorkerCtx ctx wid liftE3 $ \wrk ->
+                W.submitCert @_ @s @t @k wrk wid poolId (tx, meta, wit)
+
+            liftHandler $ withWorkerCtx ctx wid throwE $ \wrk ->
```

Comment:
changing wallet metadata. now delegating with pool id. wonder if we need to check first if wallet is already not delegating. if yes, then error is thrown as we need to first quickStakePool. I have heard two versions, this one. and also that joinStakePool just changes this field. which means changes the stake pool it delegates. Another question. what to do if we are already joined to the pool and want to rejoin to the same one. I am tentatively in favor of also throwing error. Any thought/suggestion here?

Reply:
Ah! But this is actually slightly _wrong_. We haven't joined the stake pool at the very stage and aren't yet delegating. 
We have only _made a request_ about it. The delegation will only be effective in the epoch following the insertion of the certificate in the ledger. Detecting that we are delegating is a job for #899 ! So here, there's no need to update the wallet metadata, they'll be updated when the certificate is discovered.

Reply:
thanks for this remark. removed this. hope we will tackle this in #899